source/gameengine/Ketsji/KX_KetsjiEngine.cpp

```diff
@@ -349,35 +350,40 @@ bool KX_KetsjiEngine::NextFrame()
 	 * XXX render.fps is not considred anywhere.
 	 */
 
+	// In case of non-fixed framerate, we always proceed one frame.
+	int frames = 1;
+
 	double timestep = m_timescale / m_ticrate;
 	if (!(m_flags & USE_EXTERNAL_CLOCK)) {
 		double current_time = m_clock.GetTimeSecond();
 		double dt = current_time - m_previousRealTime;
 		m_previousRealTime = current_time;
 		m_clockTime += dt * m_timescale;
 
-		if (!(m_flags & FIXED_FRAMERATE)) {
-			timestep = dt * m_timescale;
+		const double deltatime = m_clockTime - m_frameTime;
+		if (deltatime < 0.0) {
+			// We got here too quickly, which means there is nothing to do, just return and don't render.
+			// Not sure if this is the best fix, but it seems to stop the jumping framerate issue (#33088)
+			return false;
 		}
-	}
 
-	double deltatime = m_clockTime - m_frameTime;
-	if (deltatime < 0.0) {
-		// We got here too quickly, which means there is nothing to do, just return and don't render.
-		// Not sure if this is the best fix, but it seems to stop the jumping framerate issue (#33088)
-		return false;
-	}
+		// Compute the number of logic frames to do each update in case of fixed framerate.
+		if (m_flags & FIXED_FRAMERATE) {
+			const double scale = m_ticrate / m_timescale + 1e-6;
+			frames = int(deltatime * scale);
 
-	// In case of non-fixed framerate, we always proceed one frame.
-	int frames = 1;
-
-	// Compute the number of logic frames to do each update in case of fixed framerate.
-	if (m_flags & FIXED_FRAMERATE) {
-		frames = int(deltatime * m_ticrate / m_timescale + 1e-6);
+			// If the elapsed time induce a higher framerate, sleep until the next frame time point.
+			if (frames == 0) {
+				const double sleeptime = 1.0 / scale - deltatime;
+				std::this_thread::sleep_for(std::chrono::nanoseconds((long)(sleeptime * 1.0e9)));
+				frames = 1;
+			}
+		}
+		else {
+			timestep = dt * m_timescale;
```

Comment:
Just a remark, since the frame variable declaration is a bit far away, wouldn't it be better to assign the frame=1 just here below? Only to make more readable but I'm ok with your solution too